### ISSUE TITLE
Profile GPU memory usage and identify peak allocation sources

### ISSUE THREAD

**Issue by zeinali72** [zeinali72/swe-pinn#106]
## Context

Training large PINN models (width=1024, depth=6) with large PDE batches (68k+ points) can exhaust GPU memory, especially with the double autodiff through Jacobians. There is currently no visibility into memory usage patterns.

## Related Issues

- **Depends on**: #104 (profiling infrastructure)
- **Parallel-safe** with: #105 (PDE Jacobian profiling), #106 (data transfer profiling)

## Memory Components to Measure

| Component | Expected Size | Notes |
|-----------|--------------|-------|
| Model parameters | width² × depth × 4 bytes | FourierPINN: +ff_dims layer |
| Optimizer state (Adam) | 2× parameters | First and second moment |
| Epoch batch data | num_batches × batch_size × features × 4 | Pre-allocated per epoch |
| Jacobian temporaries | batch_size × 3 × 3 × 4 | Per PDE batch during forward |
| Backward pass tape | ??? | JAX internal, depends on model |
| Validation data | n_val_points × features × 4 | Loaded into GPU memory |

### 1. Measure baseline memory after model init

\`\`\`python
model, params = init_model_from_config(cfg)
opt_state = optimiser.init(params)
log_memory_usage("after_init")
\`\`\`

### 2. Measure peak memory during training step

\`\`\`python
# Use jax.profiler.save_device_memory_profile()
jax.profiler.save_device_memory_profile("memory_before_step.prof")
(params, opt_state), _ = lax.scan(scan_body, (params, opt_state), scan_inputs)
jax.block_until_ready(params)
jax.profiler.save_device_memory_profile("memory_after_step.prof")
\`\`\`

### 3. Scale test: memory vs batch_size and model_width

Create a grid:
- batch_size: [256, 512, 1024, 2048]
- model_width: [128, 256, 512, 1024]
- model_depth: [3, 4, 5, 6]

For each combination, measure peak GPU memory and flag OOM risks.

### 4. Measure validation data memory impact

Large validation datasets (42k+ points) are loaded as numpy arrays and converted to JAX arrays. Measure the memory delta.

## Files to Profile

- \`src/training/step.py\` — peak memory during gradient computation
- \`src/losses/pde.py\` — Jacobian temporary allocations
- \`src/training/data_loading.py\` — validation/training data loading (currently NOT memory-mapped)

## Acceptance Criteria

- [ ] Memory breakdown by component (params, optimizer, data, Jacobians, backward tape)
- [ ] Peak memory measured for representative configurations
- [ ] Scale test results: memory vs batch_size × model_width × depth
- [ ] Recommendations for maximum batch_size at each model scale
- [ ] Results documented in \`profiling/memory_benchmark.md\`
